### ISSUE TITLE
fullscreenchange should be dispatched on each ancestor shadow root

### ISSUE THREAD

**Issue by rniwa** [WICG/webcomponents#614]
As things stand, [requesting a fullscreen](https://fullscreen.spec.whatwg.org/#dom-element-requestfullscreen) only fires `fullscreenchange` event on each "ancestor” documents.

This would mean that any component that contains an element that can fullscreen, it needs to listen to  `fullscreenchange` event on a document. That’s pretty clumsy.

I can think of two options here:
 1. Make `fullscreenchange` both composed and bubble, and dispatch it from the fullscreen’ed element and their ancestor document instead of just on ancestor documents.  Of course, if an iframe appears within another shadow tree, we would fire an event at the iframe and bubble up, composed, as well.
 2. In addition to ancestor documents, dispatch the event on every ancestor shadow root.

I mildy prefer (1) just because WebKit’s `webkitfullscreenchange` event is fired on the fullscreen’ed element as is.  However, (1) wouldn’t work for `fullscreenerror` event.  This is probably why `fullscreenchange` is spec’ed the way it is now.

**Comment by annevk** [comment]
Why exactly is it clumsy to listen for events on the document?

@foolip @TakayoshiKochi thoughts?

**Comment by rniwa** [comment]
@annevk : It would mean that if you had 100 shadow trees, you may end up with 100 event listeners on document coming from each shadow tree to if one of the shadow tree’s element got fullscreen’ed or not.  It would be much cleaner if each shadow root had its own event listener which listens to `fullscreenchange` event that happens in its own shadow tree.

**Comment by TakayoshiKochi** [comment]
I had the same reaction originally, as @annevk.
Okay, so assuming we have a YouTube search app that shows 100 thumbnails, and each can go fullscreen. Each component wants to listen to `fullscreenchange` event, but currently if one goes fullscreen, all 100 components are notified, which is quite wasteful.

I think option 1 works, but also would like to hear from @foolip the original design decision why we get an event on document.

**Comment by annevk** [comment]
So the problem is that the Fullscreen API expects global reasoning, whereas shadow trees would like more local reasoning. Hmm.

We would need to continue to dispatch to document, but also dispatching more locally is possible. Though if we go with shadow roots that would have to happen recursively as they nest? Probably from innermost to outermost?

**Comment by rniwa** [comment]
Yeah, we’d need to do the same thing as nesting iframes.

**Comment by foolip** [comment]
Would this resolve itself if we dispatch the fullscreenchange event on the element which requested fullscreen, and have that bubble up through shadow trees? This is how webkitfullscreenchange works. There's some not-beautiful stuff that deals with the element being removed or moving between documents, but it's not so ugly it couldn't be spec'd and tested I think.

**Comment by annevk** [comment]
If we go down that route @upsuper should probably be involved in the conversation here.

**Comment by foolip** [comment]
Indeed, Gecko would be taking on some compat risk by trying to do that, and probably switching over the mozfullscreenchange event to that behavior would seem like a bad use of time.

**Comment by upsuper** [comment]
I don't quite have idea why Gecko would take compat risk by trying to switch the behavior to something similar to what WebKit and Blink currently do for webkitfullscreenchange. And I don't understand why (1) doesn't work for fullscreenerror either.

The fullscreenchange event for entering fullscreen is probably not too hard. The tricky thing is that for exiting fullscreen... yeah, if the element is removed from the document, we need to find something to dispatch the event on... and also if we change the document state synchronously but dispatch event asynchronously (as proposed in whatwg/fullscreen#74), the document tree may be mutated just before the event dispatch as well.

I don't have strong opinion here. I guess I'm fine with either... (Perhaps I miss some difficulty... I don't know...)

**Comment by foolip** [comment]
Might there not be sites that assume that the event target for mozfullscreenchange is a document?

**Comment by upsuper** [comment]
Oh, that's a good question... webkitfullscreenchange's target is an element?

I guess people usually try to handle these events together, so that's probably not a big deal, but who knows... Websites are always good at surprising browser developers :)

**Comment by foolip** [comment]
Yes, webkitfullscreenchange fires on an element and bubbles, except in some funny cases where there is no element. I think this could be spec'd in some mostly reasonable way, but as you say the web is good at surprising browser developers.

Until we decide to try to spec the prefixed Fullscreen APIs I suppose we'd all leave our prefixed APIs alone as much as possible, just trying to think through the fallout.

Assuming that we *did* make that change to Fullscreen, would it solve the problem in this issue?

**Comment by rniwa** [comment]
> if the element is removed from the document, we need to find something to dispatch the event on... and also if we change the document state synchronously but dispatch event asynchronously (as proposed in whatwg/fullscreen#74), the document tree may be mutated just before the event dispatch as well.

This is why I said `fullscreenerror` probably doesn't work well if it's dispatched on the fullscreen'ed element.

**Comment by foolip** [comment]
This is actually a problem even without https://github.com/whatwg/fullscreen/issues/74, since the `fullscreenchange` event per spec is fired at animation frame timing.

WebKit has code to fire a second event if the element is removed, and in my last attempt to match the spec in Blink I ended up changing `webkitfullscreenchange` to check if the element is still [connected](https://dom.spec.whatwg.org/#connected) to the correct document, and if not using the correct document as the event target instead. I took some inspiration from https://w3c.github.io/webappsec-csp/#report-violation where the event target depends on the circumstances.

I think we could make it work if we wanted to. So far it's seem like a hassle, but if we could fix a problem like this and alias the prefixed event, it looks more worthwhile.
